lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n if n == 1
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
```

Comment:
The line `s[1..-1]` will create an entire new array, so this is actually O(n^2) in space complexity.  Since it copies the array this makes it O(n^2) in time complexity as well.